IIIF canvas dimension task

mspalti opened the pull request:
## References
* https://github.com/DSpace/dspace-angular/issues/1396 (the issue should have been created for dspace, not dspace-angular)

## Description
Adds a new DSpace cli task for creating canvas dimension bitstream metadata.  It adds metadata to image bitstreams that belong to an IIIIF enabled item and are found in an IIIF bundle. Requires an identifier for a community, collection or item as a starting point. Additional flags for force processing, maximum items to process, skip list, and quiet mode.

Example:
```
./dspace iiif-canvas-dimensions -e user@email.org -i 1086306d-8a51-43c3-98b9-c3b00f49105f -t COLLECTION
```

## Instructions for Reviewers

List of changes in this PR:
* Updated launcher.xml
* "iiif-processing" spring xml config
* New IIIF utils class (shared with dspace-iiif package)
* New "canvasdimension" package for processing
* Integration tests


## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mspalti committed:
Added collection metadata check for iiif enabled.

mspalti committed:
Added collection metadata check for iiif enabled.

mspalti committed:
Initial work on bitstream IIIF processor.

mspalti committed:
Initial work on bitstream IIIF processor.

mspalti committed:
Added more options.

mspalti committed:
Added more options.

mspalti committed:
Added added service for image formats not supported by ImageIO (jp2)

mspalti committed:
Added added service for image formats not supported by ImageIO (jp2)

mspalti committed:
Various updates plus IT.

mspalti committed:
Refactored to use dspace-api utility class shared with CLI tool.

mspalti committed:
Moved commit to CLI class.

mspalti committed:
IT update.

mspalti committed:
Merge branch 'image-dimension-task' of https://github.com/mspalti/DSpace into image-dimension-task

mspalti committed:
Multiple updates to configs, processing classes and shared utils.

abollini commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java:
```diff
@@ -0,0 +1,219 @@
+     */
+    private boolean processBitstream(Context context, Bitstream bitstream) throws Exception {
+        boolean processed = false;
+        boolean isUnsupported = bitstream.getFormat(context).getMIMEType().contains("image/jp2");
```
it would be better imho to try to read any image and catch eventually the exceptions/null results

abollini commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java:
```diff
@@ -0,0 +1,219 @@
+                if (isUnsupported) {
+                    dims = iiifApiQuery.getImageDimensions(bitstream);
+                } else {
+                    InputStream stream = bitstreamService.retrieve(context, bitstream);
+                    dims = ImageDimensionReader.getImageDimensions(stream);
```
the inputstream must be closed in a finally block

abollini commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java:
```diff
@@ -0,0 +1,219 @@
+        dSpaceObjectService.clearMetadata(context, bitstream, "iiif",
+                "image", "width", Item.ANY);
+        dSpaceObjectService.setMetadataSingleValue(context, bitstream, "iiif",
+                "image", "width", Item.ANY, String.valueOf(dims[0]));
```
metadata should be added with `null` for the language, it is fine to clear them with ANY to cleanup manually entered information

abollini commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:
```diff
@@ -0,0 +1,203 @@
+
+
+        boolean iiifEnabled = configurationService.getBooleanProperty("iiif.enabled");
+        if (!iiifEnabled) {
+            System.out.println("IIIF is not enabled on this DSpace server.");
+            System.exit(0);
+        }
```
this is not strictly needed. The script has a value also when iiif is not enabled because it adds technical metadata at the bitstream level. It is true that without IIIF the API service will be unable to return any information but this is just the fallback for the scenario where the ImageIO class is not able to calculate it.
I would suggest to just print a warning message instead than refuse to run the script

abollini commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+            arr[1] = parent.get("height").asInt();
+            return checkDimensions(arr);
+        } catch (IOException e) {
+            e.printStackTrace();
```
as in https://github.com/DSpace/DSpace/pull/8047/files#r783792920
this must be logged

abollini commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+            while ((inputLine = in.readLine()) != null) {
+                response.append(inputLine);
+            }
+            in.close();
```
as in https://github.com/DSpace/DSpace/pull/8047/files#r783791480
the input stream should be closed in a finally block to avoid connection leaks in case of exception

abollini commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    private int[] getIiifImageDimensions(Bitstream bitstream) {
+        int[] arr = new int[2];
+        String path = iiifImageServer + bitstream.getID() + "/info.json";
```
as in https://github.com/DSpace/DSpace/pull/8047/files#r783791480
would be possible to delegate the generation of the info.json url to the `ImageContentService` ?

mspalti replied:
I added the method to `org.dspace.iiif.IIIFSharedUtils`.  On the IIIF API side we can use this method from within `ImageContentService`.

abollini reviewed (changes requested):
Hi @mspalti thanks for this valuable contribution.
It is indeed a feature that was provided by the 4Science's IIIF Addon that is very appreciated by customer. It is a pity that we were not aware of your intention to work on it otherwise we would have considered the option to donate what we already have using it as starting point to work together and improve the solution.  That said, your solution looks great I have included some suggestion inline to further improve it according to our usage experience

mspalti commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java:
```diff
@@ -0,0 +1,219 @@
+     */
+    private boolean setBitstreamMetadata(Context context, Bitstream bitstream, int[] dims) throws Exception {
+        dSpaceObjectService.clearMetadata(context, bitstream, "iiif",
+                "image", "width", Item.ANY);
+        dSpaceObjectService.setMetadataSingleValue(context, bitstream, "iiif",
+                "image", "width", Item.ANY, String.valueOf(dims[0]));
+        dSpaceObjectService.clearMetadata(context, bitstream, "iiif",
+                "image", "height", Item.ANY);
+        dSpaceObjectService.setMetadataSingleValue(context, bitstream, "iiif",
+                "image", "height", Item.ANY, String.valueOf(dims[1]));
```
Hi @abollini , so `IIIFUtils` and `ImageContentService` are part of the spring boot web application. Since we already have an `IIIFSharedUtils` class in `org.dspace.iiif` it might make sense for both the cli process and web application to use the shared utils for constants and the `info.json` path. I'll refactor accordingly. Let me know if you see a better approach.

mspalti commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/Util.java:
```diff
@@ -0,0 +1,34 @@
+     * @return
+     */
+    public static int[] checkDimensions(int[] dims) {
+        if (dims[0] < 1200 || dims[1] < 1200) {
```
It's poorly worded and  a bit confusing but the comment I copied from the IIIF specification says less than 1200 pixels on either edge, which seems to say we need ` ||`. I will double check the spec.

mspalti committed:
Multiple updates.

mspalti commented:
@abollini , I wish we had started working together on this one, too. I knew that I had jumped into work that might be helped along by more discussion in the community and 4Science input. There were just some things that really needed to be done within the timeline for 7.2 and that's the reason for the quick start. Thanks for your willingness to collaborate!

mspalti committed:
Import shared constants into IIIFUtils.

mspalti committed:
Refactored util package

mspalti commented:
@abollini , if you approve of the changes I made here, I will address your similar suggestions for #8015 after this PR is merged. They can both rely on the new `IIIFSharedUtils`

tdonohue commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:
```diff
@@ -0,0 +1,202 @@
+        if (line.hasOption('i')) {
+            identifier = line.getOptionValue('i');
+        } else {
+            System.out.println("An identifier for a Community, Collection, or Item must be provided.");
```
Can we print out the help text here as well?  The first time I ran this with no arguments, I expected it to error out, but also tell me what a valid argument would look like.  

There's also no "--help" or "-h" option...so, it's impossible to see the list of valid arguments from commandline.  We should add a "-h" option

tdonohue commented on dspace/config/launcher.xml:
```diff
@@ -366,4 +366,11 @@
     </command>
+    <command>
+        <name>canvas-dimensions</name>
+        <description>Add canvas width and height metadata for IIIF enabled items.</description>
```
We may want to improve this description, as I first understood it to say that it adds the canvas height/width __to the Item__. But, it's obviously adding it to the Bitstream.  Maybe say something like this?

"Automatically add the IIIF canvas width/height metadata to Bitstreams of IIIF enabled Items".

tdonohue commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:
```diff
@@ -0,0 +1,202 @@
+                processed = 1;
+                break;
+            default:
+                break;
```
We might want to log an error of some sort if an unsupported object type is passed in. Otherwise, it appears this will just "do nothing" and be quiet about it.

tdonohue commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:
```diff
@@ -0,0 +1,202 @@
+                break;
+        }
+        // commit changes
+        context.commit();
```
We should only commit changes if something is updated.  this should at least check that >=1 items were processed

tdonohue reviewed (changes requested):
@mspalti : I tested this today & also gave it a quick code review. With regards to testing, it mostly works, but it's a bit hard to use for a few reasons:

1. UUIDs don't appear to work.  So, an argument `./dspace canvas-dimensions -i [uuid] -e [email]` results in an error: `Cannot resolve [uuid] to a DSpace object`.  Only Handles work.  Ideally either UUIDs or handles would work if possible.
2. The CLI command is missing help information, i.e. an `-h` option should exist, otherwise you cannot figure out what commands are valid without looking at documentation (or code), see more inline below.

Some other comments inline regarding the code structure and other minor suggestions.

mspalti commented:
@tdonohue , I think that using UUID as the identifier means we require an additional parameter that specifies the dso type (Community, Collection, Item). That's what I see in other scripts, like `ItemExportCLITool`.  Basically we need to know which service to call so we can do the UUID lookup. I'm adding this now.

This script currently works like media filter, which unless I'm mistaken only works with handles. I'm a little surprised if that's still the case, though, so perhaps I missed something.

mspalti committed:
Multiple updates.

mspalti commented on dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:
```diff
@@ -0,0 +1,202 @@
+ *
+ * @author Michael Spalti mspalti@willamette.edu
+ */
+public class CanvasDimensionCLI {
```
I agree that this should eventually be updated. Perhaps an issue for 7.3

abollini replied:
CurationTask is also something to consider as it can be run as  part of the workflow. We have it in our addon so it is something that we can eventually provide that for 7.3+

mspalti commented:
I think I've addressed all feedback now. The script works with both UUID and handle. If the UUID is provided then an extra dso type parameter is also required so the script knows which service to use when retrieving the dso. e.g.

```
iiif-canvas-dimensions -e user@email.org -i 1086306d-8a51-43c3-98b9-c3b00f49105f -t COLLECTION
```

Example provided in `--help`.

tdonohue commented:
@mspalti : Apologies, I missed your comment yesterday about UUIDs.  It is possible to resolve an object by ONLY the UUID (no type is necessary) if you use the DSpaceObjectUtils.findDSpaceObject(context, uuid) method: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/app/util/DSpaceObjectUtilsImpl.java#L37   This is the same manner that our REST API uses to resolve objects based on UUIDs (as UUIDs are unique) when you call the `/api/dso/find?uuid=` endpoint: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/UUIDLookupRestController.java#L93

That said, as noted in that method, it's not the most performant "lookup" as we literally just try to lookup that UUID for each type of object.  So, while requiring the type is not actually required, it probably will give slightly better performance. 

In this scenario, as you are only doing one object lookup, we probably could drop the requirement for a type...but, honestly, I'm OK either way.  So, your approach in this PR is fine.

I haven't had a chance to give this an updated review, but it does look (at a quick glance) that you've addressed my major concerns.  Will try and get back to this later today if possible.

mspalti commented:
@tdonohue , that's good to know. It thought there might be such a method. So I think it would be better to drop the "type" parameter. I won't be able to get to that today but it should be a quick fix tomorrow.

mspalti committed:
Modified CLI and renamed packages.

mspalti committed:
Merge branch 'main' into image-dimension-task

mspalti commented:
OK. This is now ready for another review. I think I've addressed all concerns. Sorry this took a while I'm on the road with a new laptop and bad wifi.

abollini commented on dspace-api/src/main/java/org/dspace/iiif/IIIFApiQueryServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    public int[] getImageDimensions(Bitstream bitstream) {
+        return getIiifImageDimensions(bitstream);
+    }
+
+    /**
+     * Retrieves image dimensions from the image server (IIIF Image API v.2.1.1).
+     * @param bitstream the bitstream DSO
+     * @return image dimensions
+     */
+    private int[] getIiifImageDimensions(Bitstream bitstream) {
```
should we just retain the public method?

abollini reviewed (approved):
changes look good to me. @tdonohue I have reviewed also your question flagging them as resolved where appropriate to speedup your second review round.

tdonohue reviewed (approved):
👍 Reviewed/tested again today. Thanks @mspalti !  This all looks good to me now.  I've verified all my prior feedback was addressed, and re-tested and the commandline tool works well (also works for just UUIDs now).
